CONTRIBUTING.md

```diff
@@ -144,6 +144,22 @@ Note that we usually also practice top-down ordering here; where these are in
 conflict, make a choice that you think makes sense. For getters and setters, the
 order should typically mirror the order of the fields in the type definition.
 
+#### Attribute ordering
+
+Order attributes so that documentation appears first, and the more "important"
+attributes appear last.  For example:
+
+```rust
+/// Doc comment always first
+#[cfg(feature-gates)]
+#[allow(lint-configuration)]
+#[non_exhaustive]
+#[derive(Clone, Debug)]
+pub struct Foo;
+```
```

Comment:
How would you order these as import? From most to least important, I'd suggest:

- `cfg` (can turn the entire thing off)
- `non-exhaustive` (controls important downstream behavior)
- derive stuff (trait implementations)
- lint stuff (irrelevant to downstreams)

But, I'd probably order like this:

- Docstring
- cfg
- non-exhaustive
- lints
- derive

I guess maybe because I feel like the derives should not concern themselves with the attributes/the procedural macros should be close to the syntax they use as input?

Reply:
Probably important isn't the right word. The way I think about this is how much effect each attribute has on the type. `derive` has the most effect, then `non_exhaustive` . Lints, cfg gates and comments have ~no effect on the type.